src/cliffordsynthesis/GateEncoding.cpp

```diff
@@ -0,0 +1,364 @@
+                        changes = changes && (data.z[gateStep][c] == data.z[gateStep - 1][c]);
+                    }
+
+                    changes = logicbase::LogicTerm::implies(data.gTwoQubit[gateStep][a][b], changes);
+                    data.lb->assertFormula(changes);
+                }
+                ++bIt;
+            }
+            ++aIt;
+        }
+    }
+}
+void cs::GateEncoding::makeMultiGateEncoding(const SynthesisData& data) {
+    auto changes = logicbase::LogicTerm(true);
+    // CONSISTENCY
+    // One gate per qubit, per step
+    for (unsigned int gateStep = 1; gateStep < data.timesteps + 1; ++gateStep) {
+        for (unsigned int a = 0; a < data.nqubits; ++a) {
+            std::vector<logicbase::LogicTerm> vars{};
+            for (auto gate: Gates::SINGLE_QUBIT) {
+                vars.emplace_back(data.gS[gateStep][Gates::toIndex(gate)][a]);
+            }
+            for (unsigned int b = 0; b < data.nqubits; ++b) {
+                if (a == b) {
+                    continue;
+                }
+                vars.emplace_back(data.gTwoQubit[gateStep][a][b]);
+                vars.emplace_back(data.gTwoQubit[gateStep][b][a]);
+            }
+            data.lb->assertFormula(encodings::exactlyOneCmdr(
+                    encodings::groupVars(vars, static_cast<std::size_t>(vars.size() / 2)),
+                    logicbase::LogicTerm::noneTerm(), data.lb.get()));
+        }
+    }
+    // Maximum any combination of 1 and 2 qubit gates adding up to n
+    for (unsigned int gateStep = 1; gateStep < data.timesteps + 1; ++gateStep) {
+        changes = logicbase::LogicTerm(0);
+        for (unsigned int a = 0; a < data.nqubits; ++a) {
+            for (auto gate: Gates::SINGLE_QUBIT) {
+                changes = changes + data.gS[gateStep][Gates::toIndex(gate)][a];
+            }
+            for (unsigned int b = 0; b < data.nqubits; ++b) {
+                if (a == b) {
+                    continue;
+                }
+                changes = changes + data.gTwoQubit[gateStep][a][b] + data.gTwoQubit[gateStep][a][b];
+            }
+        }
+        changes = changes < logicbase::LogicTerm(static_cast<int>(data.nqubits + 1));
+        data.lb->assertFormula(changes);
+    }
+
+    // GATE CONSTRAINTS
+    for (unsigned int gateStep = 1; gateStep < data.timesteps + 1; ++gateStep) {
+        logicbase::LogicTerm rChanges = data.r[gateStep - 1];
+        auto                 aIt      = data.qubitChoice.cbegin();
+        for (unsigned int a = 0; a < data.nqubits; ++a) {
+            // NO GATE
+            changes = logicbase::LogicTerm(true);
+            changes = changes && (data.x[gateStep][a] == data.x[gateStep - 1][a]);
+            changes = changes && (data.z[gateStep][a] == data.z[gateStep - 1][a]);
+            changes = logicbase::LogicTerm::implies(data.gS[gateStep][0][a], changes);
+            data.lb->assertFormula(changes);
```

Comment:
these seem rather redundant with the constraints above. lots of this duplication can be saved by extracting the common functionality into functions.